File: data/website.go
package data

import (
	"fmt"
)

type WebsiteResponseData struct {
	Metadata      Metadata      `json:"metadata"`
	Blog          Blog          `json:"blog"`
	QuoteOfTheDay QuoteOfTheDay `json:"quoteOfTheDay"`
}

type Metadata struct {
	Url         string   `json:"url"`
	Title       string   `json:"title"`
	Description []string `json:"description"`
	Socials     []Social `json:"socials"`
}

type Social struct {
	Name string `json:"name"`
	Href string `json:"href"`
}

type Blog struct {
	Posts       []BlogPost `json:"posts"`
	Description string     `json:"description"`
}

type BlogPost struct {
	Title   string `json:"title"`
	Summary string `json:"summary"`
	Slug    string `json:"slug"`
	PubDate string `json:"publishedAt"`
}

type QuoteOfTheDay struct {
	Text   string `json:"text"`
	Author string `json:"author"`
}

type Website struct {
	Title         string
	Description   []string
	Url           string
	Socials       []Social
	BlogPosts     []BlogPost
	QuoteOfTheDay QuoteOfTheDay
}

func GetWebsiteData() Website {
	url := "https://www.apl.directory/api"

	response, err := MakeRequest(url, &WebsiteResponseData{})

	if err != nil {
		fmt.Println("Error fetching website data:", err)
		return Website{}
	}
	data := response.(*WebsiteResponseData)

	posts := data.Blog.Posts

	return Website{
		Title:       data.Metadata.Title,
		Description: data.Metadata.Description,
		Url:         data.Metadata.Url,
		Socials:     data.Metadata.Socials,
		BlogPosts:   posts,
		QuoteOfTheDay: QuoteOfTheDay{
			Text:   data.QuoteOfTheDay.Text,
			Author: data.QuoteOfTheDay.Author,
		},
	}
}

File: Sources/apl9000/Website.swift
import Foundation

// Shape of the https://www.apl.directory/api response.
private struct WebsiteResponseData: Decodable {
    let metadata: Metadata
    let blog: Blog
    let quoteOfTheDay: QuoteOfTheDay
}

private struct Metadata: Decodable {
    let url: String
    let title: String
    let description: [String]
    let socials: [Social]
}

private struct Blog: Decodable {
    let posts: [BlogPost]
}

struct Social: Decodable {
    let name: String
    let href: String
}

struct BlogPost: Decodable {
    let title: String
    let summary: String
    let slug: String
    let publishedAt: String
}

struct QuoteOfTheDay: Decodable {
    let text: String
    let author: String

    static let empty = QuoteOfTheDay(text: "", author: "")
}

struct Website {
    let title: String
    let description: [String]
    let url: String
    let socials: [Social]
    let blogPosts: [BlogPost]
    let quoteOfTheDay: QuoteOfTheDay

    static let empty = Website(
        title: "",
        description: [],
        url: "",
        socials: [],
        blogPosts: [],
        quoteOfTheDay: .empty
    )
}

func getWebsiteData() async -> Website {
    let url = "https://www.apl.directory/api"

    guard let data = await fetchJSON(url, as: WebsiteResponseData.self) else {
        return .empty
    }

    return Website(
        title: data.metadata.title,
        description: data.metadata.description,
        url: data.metadata.url,
        socials: data.metadata.socials,
        blogPosts: data.blog.posts,
        quoteOfTheDay: data.quoteOfTheDay
    )
}
